### PR TITLE
fix(api): various window-related function fixes

### PR DESCRIPTION
**Pull request by seandewar**
Follows up on some issues in #25550 and (a bunch of) other things

Also includes Vim patches to issues I found while working on things: 8.2.3862, 9.1.{0116,0118,0119,0121,0128,0130} (and 9.1.0117 as a partial patch).

### CODE REVIEW

**Comment by willothy** [comment]
This looks great! I'm working on fixes for some of the [split behavior](https://github.com/neovim/neovim/pull/25550#discussion_r1474165527) btw, unless you were planning on doing that.

File: src/nvim/api/win_config.c
Comment (willothy) [inline src/nvim/api/win_config.c:498]: The only uses of `win_split_flags` add `| WSP_NOENTER`. Maybe we should just move that into the function?

Comment (seandewar) [reply src/nvim/api/win_config.c:498]: Subjective (especially as `win_split_flags` is static too), but I think I prefer `WSP_NOENTER` being set at the caller's discretion.

**Comment by seandewar** [comment]
> This looks great! I'm working on fixes for some of the [split behavior](https://github.com/neovim/neovim/pull/25550#discussion_r1474165527) btw, unless you were planning on doing that.

Feel free to work on that! I only plan to address some random things I find here.

**Comment by willothy** [comment]
> Feel free to work on that! I only plan to address some random things I find here.

Sounds good!

**Comment by seandewar** [comment]
TODO: (partially, 'cuz WinNewPre, _sigh_) port the following patches I've upstreamed to address the WIP in the first commit:

https://github.com/vim/vim/commit/0fd44a5ad81ade342cb54d8984965bdedd2272c8
https://github.com/vim/vim/commit/96cc4aef3d47d0fd70e68908af3d48a0dce8ea70
https://github.com/vim/vim/commit/704966c2545897dfcf426dd9ef946aeb6fa80c38
https://github.com/vim/vim/commit/f865895c874b0936b0563ebfef7490aac8cb8a1f

**Comment by seandewar** [comment]
About time I open this for reviews before the scope increases more :see_no_evil:.

There are a few other things I discovered that I haven't addressed here:

- `tabpagewinnr()` and `nvim_tabpage_get_win()` don't work for non-current tabpages with `tp_curwin`s pointing to external windows (the latter calls `abort()`). This is a consequence of how "tabpage independence" of external windows are implemented (they're moved between tabpages when entering another).

- Unlike the minimum value of `&lines` which is based on the minimum frame height, the minimum value of `&columns` is not based on the minimum frame width. This can cause the assertion `(size_t)grid_line_maxcol <= linebuf_size` in `grid_line_start()` to fail (and maybe other issues) if the frames can't fit:

  `:execute 'vsplit|'->repeat(99) | set columns=12`

- Some of the textlocky API functions probably ought to check `curbuf_locked()` too...

File: src/nvim/window.c
Comment (seandewar) [inline src/nvim/window.c:3367]: `last_status()` only works for the current tabpage, and is not called for other tabpages when `&laststatus` changes (at least not until they're entered via `enter_tabpage()`).

In the context of `winframe_restore()`, this means if `wp` is in another tabpage, and say, `&laststatus` changed from 3 (global statusline) to something else before it's entered, this may add a statusline to a window that already has a horizontal separator (or vice versa if `&laststatus` is set the other way around).

This doesn't seem to have any observable effect that I can find though, as the frame sizes aren't changed by this, and `last_status()` will fix the hseps/statuslines upon entering it anyway.

File: src/nvim/eval/window.c
Comment (seandewar) [inline src/nvim/eval/window.c:705]: `check_split_disallowed(targetwin)` ought to be `check_split_disallowed(wp)` instead, but I'm not convinced checking `b_split_locked` is needed when split-moving anyway; we're not producing new views into a closing buffer by doing it.

I'll probably end up addressing this and others sometime after vim/vim#14109.

File: src/nvim/window.c
Comment (seandewar) [inline src/nvim/window.c:1945]: `win_comp_pos()` probably isn't needed here after the commit which added `frame_comp_pos()` to `winframe_restore()`; at the very least, I didn't add it to `nvim_win_set_config()`...

I believe the reason it's used after `winframe_remove()` and `last_status()` above anyway is because `last_status()` may steal lines from other frames to make room for a missing statusline.

**Review by zeertzjq — APPROVED** [review]

**Comment by zeertzjq** [comment]
I also pushed some follow-up commits.

**Comment by seandewar** [comment]
Thanks for that! :hugs:

After sleeping on things, I've realized there's a few more issues I missed:

- `nvim_win_set_config` should select a new `tp_curwin` if `win` was moved from a different tabpage.
- `winframe_remove` takes a `tp` argument, but its flattening logic references `topframe` (_lolwut_).
- `switch_win` with `no_display` doesn't update the `topframe`, which should be done before `win_split_ins` in `nvim_win_set_config`.

None of these are regressions from this PR, but they should be addressed before merging. :see_no_evil:

**Comment by zeertzjq** [comment]
Oh, I have to move some changes out of this PR then.

**Comment by seandewar** [comment]
Its OK, I think I've fixed those issues now. (I'll push in a bit)

**Comment by zeertzjq** [comment]
I notice that moving these changes out of the PR still conflicts, so I pushed back to the previous state.

**Comment by seandewar** [comment]
Okie dokie, I've pushed the commits on top of that state (and amended the [`fix(api): win_set_config set tp_curwin of win moved from other tabpage`](https://github.com/neovim/neovim/pull/27330/commits/b52d15853e89149472c1ecd9cce3a84e4af0785a) one to add a bit more to the test).

Here's the [diff](https://github.com/neovim/neovim/compare/60e978539bc5ff66aacb023528088cfb89cf5559..c3d22d32ee4b4c1911ec15f2a77683d09b09f845) of that.

**Review by zeertzjq — APPROVED** [review]

**Comment by seandewar** [comment]
Ok, I'm totally _over_ this PR at this point. :laughing:

Once checks are green, I'm happy to merge immediately if that sounds good. :+1:

File: src/nvim/eval/window.c
Comment (seandewar) [inline src/nvim/eval/window.c:965]: As mentioned in the commit, this would be better off using `{un}use_tabpage`, so `tp_curwin` gets set too. This would address the inconsistency where:

```vim
:call win_execute(w, "let curwin_nr1 = tabpagewinnr(1)")
:let curwin_nr2 = tabpagewinnr(1)
```

Here it's possible for curwin_nr1 != curwin_nr2 if these commands are run from
the 1st tabpage, but window "w" is in the 2nd (as the 1st tabpage's tp_curwin
may still be invalid). I'll probably PR a fix for that later in Vim.

**Comment by willothy** [comment]
Awesome! Not sure I could have provided any helpful reviews here as a lot of this is beyond my Vim knowledge (for now...) but this PR has been very informative to read through.

**Comment by seandewar** [comment]
Now time to hope I didn't break anything :innocent:

**Pull request merged**
